3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,311 @@
+    new PiiStrategyPseudonymizeSerializer
+
+  def transformer(event: EnrichedEvent): Unit = {
+    val modifiedFields: ModifiedFields = fieldList.flatMap(_.transform(event, strategy))
+    event.pii =
+      if (emitIdentificationEvent && modifiedFields.nonEmpty) write(PiiModifiedFields(modifiedFields, strategy))
+      else null
+  }
+}
+
+/**
+ * Specifies a scalar field in POJO and the strategy that should be applied to it.
+ * @param fieldMutator the field mutator where the strategy will be applied
+ */
+final case class PiiScalar(fieldMutator: Mutator) extends PiiField {
+  override def applyStrategy(fieldValue: String, strategy: PiiStrategy): (String, ModifiedFields) =
+    if (fieldValue != null) {
+      val modifiedValue = strategy.scramble(fieldValue)
+      (modifiedValue, List(ScalarModifiedField(fieldMutator.fieldName, fieldValue, modifiedValue)))
+    } else (null, List())
+}
+
+/**
+ * Specifies a strategy to use, a field mutator where the JSON can be found in the EnrichedEvent POJO, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a JSON path within the contexts where this strategy will
+ * be applied (the path may correspond to multiple fields).
+ *
+ * @param fieldMutator the field mutator for the JSON field
+ * @param schemaCriterion the schema for which the strategy will be applied
+ * @param jsonPath the path where the strategy will be applied
+ */
+final case class PiiJson(fieldMutator: Mutator, schemaCriterion: SchemaCriterion, jsonPath: String) extends PiiField {
+  implicit val json4sFormats = DefaultFormats
+
+  override def applyStrategy(fieldValue: String, strategy: PiiStrategy): (String, ModifiedFields) = {
+    val modifiedFields = MutableList[JsonModifiedField]()
+    if (fieldValue != null) {
+      val parsedAndSubistuted = parse(fieldValue) match {
+        case JObject(jObject) =>
+          val jObjectMap = jObject.toMap
+          val updated = jObjectMap.filterKeys(_ == "data").mapValues {
+            case JArray(contexts) =>
+              JArray(contexts.map {
+                case JObject(context) => modifyAndGetValues(context, strategy, modifiedFields)
+                case x                => x
+              })
+            case JObject(unstructEvent) => modifyAndGetValues(unstructEvent, strategy, modifiedFields)
+            case x                      => x
+          }
+          JObject((jObjectMap ++ updated).toList)
+        case x => x
+      }
+      val rendered  = render(parsedAndSubistuted)
+      val compacted = compact(rendered)
+      (compacted, modifiedFields.toList)
+    } else (null, modifiedFields.toList)
+  }
+
+  /**
+   * Modifies the object if applicable and adds to the list of modified values, if applicable.
+   */
+  private def modifyAndGetValues(context: List[(String, json4s.JValue)],
+                                 strategy: PiiStrategy,
+                                 modifiedFields: MutableList[JsonModifiedField]): JObject = {
```

Comment:
can't this function return `listOfModifiedValues` instead of passing and mutating `modifiedFields`?

Reply:
Done (good point although readability is worse now if you ask me)